Add support for `async` polling

leighleighleigh opened the pull request:
This PR adds a few useful feature flags, resulting from my own usage with the [atsamd](https://github.com/atsamd-rs/atsamd) HAL, which has a mix of `embedded-hal` `v0.2.7` and `v1.0.0` scattered throughout.


<details>
<summary>Original PR text, relevant to first commit only.</summary>

At a high level, the `eh1` and `eh0` features simply change the trait requirement for the encoder pins - using `InputPin`, but from different versions of the `embedded-hal` crate. `eh1` uses v1.0.0, `eh0` uses v0.2.7.

The `async` feature adds the `Wait` trait requirement to the encoder pins, and enables a new `poll_async` function.
This function `select()`s the `wait_for_any_edge()` function on each pin, until an edge is detected, then calls the regular `poll()` function. _I haven't tested this `poll_async()` on any hardware yet - I shall do so in the next 24 hours._ 

I've added a few new dependencies, most are fairly self-explanatory. I chose to use `embassy-futures` instead of the regular `futures` crate because of it's nice `select3` function, and the crate has no further dependencies of it's own, so felt safe.

Thanks for the handy crate!

</details>

## Updated Description 
__...updated on epoch 1737691201__.

After a bit of testing in the wild, I realized that my approach to both `async` and `eh0/eh1` compatibility had a couple fatal flaws:
 1. It wasn't possible to use blocking AND async encoders within the same program, partially due to how the internal `InputPin` trait was affected by the feature flags in use. 
 2. There's really no need to provide `eh0`-compatibility 'from inside the library'. Even though it's 'ugly', I think it's better to rely on a user-wrapped `Forward<Pin>` from the `embedded-hal-compat` crate instead. I've added `rotary_eh0` and `linear_eh0` examples that demonstrates this (although there is probably no need to have both).
 3. The way I had sprinkled `#[cfg(feature = XYZ)]` around resulted in non-additive features, and it was a bit ugly.
 4. It changed the API of `poll()` to `poll_async()`, which is an admittedly a minor change, but one that is inconsistent with how async rust is usually done.
 
To address these shortcomings, I've opted to create an empty `PollMode` marker trait, akin to the `DriverMode` trait in projects such as `esp-hal`'s peripheral drivers. This allows for unique implementations of `poll()` function, depending on this mode, which can be either `Blocking` or `Async`. Common logic for each implementation (updating the `Decoder`) has been moved to a private method `update()`.

Apologies for hacking on this existing PR - it was a bit more of a WIP than I expected.

regexident commented:
@leighleighleigh thanks! I'll try to give this a review this weekend.

leighleighleigh commented:
FWIW I just tested this on real hardware, and `poll_async()` seems to be working well for me.

leighleighleigh commented:
The approach using 'wait_for_any_edge()` proved to be unreliable at high speed. I've opted to instead track the last-observed pin state in some internal variables, and then check for rising/falling edges conditionally based on those states. In my testing with a pretty low-quality encoder, I got reliable operation with the `FullStep` and `HalfStep` modes. `QuadStep` seems to skip steps occasionally, which is to be expected.

regexident commented on quadrature-encoder/Cargo.toml:
```diff
@@ -15,9 +15,19 @@ version = { workspace = true }
 num-traits = { workspace = true }
 quadrature-decoder = { version = "0.1.1", path = "../quadrature-decoder", default-features = false }
+futures = { version = "0.3.31", default-features = false, optional = true }
+embassy-futures = "0.1.1"
```
This should probably be optional as well, I think?

regexident commented on quadrature-encoder/examples/linear.rs:
Would you mind adding a separate `linear_async.rs` example?

With all the `#[cfg(…)]` noise the actual API being illustrated sort of gets lost otherwise.

regexident commented on quadrature-encoder/examples/rotary.rs:
Would you mind adding a separate `rotary_async.rs` example?

With all the `#[cfg(…)]` noise the actual API being illustrated sort of gets lost otherwise.

regexident commented on quadrature-encoder/src/encoder/incremental.rs:
```diff
@@ -38,16 +45,24 @@ where
     /// Creates an incremental encoder driver for the given pins.
-    pub fn new(pin_clk: Clk, pin_dt: Dt) -> Self
+    /// NOTE: eh1 requires mutable pin references, but eh0 does not, which upsets clippy sometimes.
+    #[allow(unused_mut)]
```
> NOTE: eh1 requires mutable pin references, but eh0 does not, which upsets clippy sometimes.

In that case we should be able to limit the `allow(…)` to just `eh0`, this limiting its scope, I think? 

```suggestion
    #[cfg_attr(not(feature = "eh1"), allow(unused_mut))]
```

regexident commented on quadrature-encoder/src/encoder/incremental.rs:
```diff
@@ -27,6 +32,8 @@ pub struct IncrementalEncoder<Mode, Clk, Dt, Steps = FullStep, T = i32> {
     pin_dt: Dt,
     is_reversed: bool,
     _mode: PhantomData<Mode>,
+    pin_clk_state: bool,
+    pin_dt_state: bool,
```
```suggestion
    pin_clk_state: bool,
    pin_dt_state: bool,
    is_reversed: bool,
    _mode: PhantomData<Mode>,
```

regexident commented on quadrature-encoder/src/encoder/incremental.rs:
```diff
@@ -38,16 +45,24 @@ where
             decoder: Default::default(),
             pin_clk,
             pin_dt,
             is_reversed: false,
             _mode: PhantomData,
+            pin_clk_state,
+            pin_dt_state,
```
```suggestion
            pin_clk_state,
            pin_dt_state,
            is_reversed: false,
            _mode: PhantomData,
```

regexident commented on quadrature-encoder/src/encoder/indexed.rs:
```diff
@@ -28,6 +33,9 @@ pub struct IndexedIncrementalEncoder<Mode, Clk, Dt, Idx, Steps = FullStep, T = i
     pin_idx: Idx,
     is_reversed: bool,
     _mode: PhantomData<Mode>,
+    pin_clk_state: bool,
+    pin_dt_state: bool,
+    pin_idx_state: bool,
```
```suggestion
    pin_clk_state: bool,
    pin_dt_state: bool,
    pin_idx_state: bool,
    is_reversed: bool,
    _mode: PhantomData<Mode>,
```

regexident commented on quadrature-encoder/src/encoder/indexed.rs:
```diff
@@ -39,18 +47,27 @@ where
             pin_clk,
             pin_dt,
             pin_idx,
             is_reversed: false,
             _mode: PhantomData,
+            pin_clk_state,
+            pin_dt_state,
+            pin_idx_state,
```
```suggestion
            pin_clk_state,
            pin_dt_state,
            pin_idx_state,
            is_reversed: false,
            _mode: PhantomData,
```

regexident commented on quadrature-encoder/src/encoder/indexed.rs:
```diff
@@ -39,18 +47,27 @@ where
-    pub fn new(pin_clk: Clk, pin_dt: Dt, pin_idx: Idx) -> Self
+    /// Creates an indexec incremental encoder driver for the given pins.
+    /// NOTE: eh1 requires mutable pin references, but eh0 does not, which upsets clippy sometimes.
+    #[allow(unused_mut)]
```
> NOTE: eh1 requires mutable pin references, but eh0 does not, which upsets clippy sometimes.

In that case we should be able to limit the `allow(…)` to just `eh0`, this limiting its scope, I think? 

```suggestion
    #[cfg_attr(not(feature = "eh1"), allow(unused_mut))]
```

regexident commented on quadrature-encoder/src/main.rs:
This was meant to be removed, I guess?

regexident commented on quadrature-encoder/src/encoder/incremental.rs:
```diff
@@ -111,6 +123,31 @@ where
         }))
     }
 
+    /// Waits asyncronously for either two pins to change state, then runs poll()
```
```suggestion
    /// Waits asynchronously for either two pins to change state, then runs poll()
```

regexident commented on quadrature-encoder/src/encoder/indexed.rs:
```diff
@@ -39,18 +47,27 @@ where
 {
-    /// Creates an incremental encoder driver for the given pins.
-    pub fn new(pin_clk: Clk, pin_dt: Dt, pin_idx: Idx) -> Self
+    /// Creates an indexec incremental encoder driver for the given pins.
```
```suggestion
    /// Creates an indexed incremental encoder driver for the given pins.
```

regexident commented on quadrature-encoder/src/traits.rs:
```diff
@@ -0,0 +1,17 @@
+#[cfg(feature = "eh1")]
+use eh1::digital::InputPin as EhalInputPin;
+
+#[cfg(feature = "eh0")]
+use eh0::digital::v2::InputPin as EhalInputPin;
```
This makes the `eh0` and `eh1` mutually exclusive, which violates the strictly additive semantics of rust features:

```terminal
cargo check --all-features
```

```plain
error[E0252]: the name `EhalInputPin` is defined multiple times
 --> quadrature-encoder/src/traits.rs:5:5
  |
2 | use eh1::digital::InputPin as EhalInputPin;
  |     -------------------------------------- previous import of the trait `EhalInputPin` here
...
5 | use eh0::digital::v2::InputPin as EhalInputPin;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `EhalInputPin` reimported here
  |
  = note: `EhalInputPin` must be defined only once in the type namespace of this module
help: you can use `as` to change the binding name of the import
  |
5 | use eh0::digital::v2::InputPin as OtherEhalInputPin;
```

regexident commented on Cargo.toml:
```diff
@@ -8,7 +8,7 @@ homepage = "https://github.com/regexident/quadrature"
 license = "MPL-2.0"
 edition = "2021"
-version = "0.1.1"
+version = "0.1.2"
```
Would you mind undoing this change? I like to have separate commits for the version-bumping, that go with the actual release.

regexident reviewed (changes requested):
Thanks for this great contribution! 🙇 

Looking at the CLI failures it looks like the next steps to get the PR ready for merging are:

2. Resolve change requests.
2. Fix formatting by running `cargo fmt`.
3. Remove stray `quadrature-encoder/src/main.rs` file.
4. Add entry to `CHANGELOG.md` file:

    ```markdown
    ## [Unreleased]

    ### Added

    - Added support for async-await (via optional `async` crate feature):
      - Added `async fn poll_async()` method to `IncrementalEncoder<…>`.
      - Added `async fn poll_async()` method to `IndexedIncrementalEncoder<…>`.
    ```

regexident commented:
@leighleighleigh lemme know if I should take it from here or if you want to do it yourself.

leighleighleigh commented:
I have updated the original description - apologies for the delay, I've been AFK.
I've incorporated most of your feedback, thank you for that, but I am yet to add a section on `async` usage to the `README.md`. I'm still happy to do that, as long you are comfortable with how I've handled the new `into_async()` approach!

I've also got a full `embassy-mock` example waiting in the wings, pending this [embedded-hal-mock](https://github.com/dbrgn/embedded-hal-mock/pull/125#issue-2808430941) PR.

regexident commented:
> I have updated the original description - apologies for the delay, I've been AFK.

No worries! I just wanted to make sure you weren't waiting on me for a release with your features.

> I've incorporated most of your feedback, thank you for that, but I am yet to add a section on `async` usage to the `README.md`. I'm still happy to do that, as long you are comfortable with how I've handled the new `into_async()` approach!

I like the `.into_async()`/`.into_blocking()` approach you took! Very ergonomic.

FYI: I've just rebased the branch onto `main` (I try to keep a linear history) and pushed a few commits, including a README section.

I'll wait for CI to pass, merge it and release a `0.2.0`.

> I've also got a full `embassy-mock` example waiting in the wings, pending this [embedded-hal-mock](https://github.com/dbrgn/embedded-hal-mock/pull/125#issue-2808430941) PR.

👨🏻‍🍳👌🏻

That looks like a nice improvement to `embedded-hal-mock` indeed. Looking forward to it being merged!

leighleighleigh commented:
Awesome, thanks!

regexident commented:
I just published `0.2.0` with your changes. 🚀 

Thanks for this great contribution! 🙏
